### PR TITLE
Update Z3SmtSoftConstraintEncoder

### PR DESCRIPTION
**Pull request by d367wang**
Add abstract methods for Z3Smt solver only.  This is only used in value-inference.

Co-authored-by: Jenny Xiang <j.tt.xiang@gmail.com>

### CODE REVIEW

File: src/checkers/inference/solver/backend/z3smt/encoder/Z3SmtSoftConstraintEncoder.java [hunk 1]
```diff
@@ -34,44 +34,63 @@ public Z3SmtSoftConstraintEncoder(
     protected void addSoftConstraint(Expr serializedConstraint, int weight) {
         softConstraints.append("(assert-soft " + serializedConstraint + " :weight " + weight + ")\n");
     }
-    
+
+
```
Comment (d367wang) [inline src/checkers/inference/solver/backend/z3smt/encoder/Z3SmtSoftConstraintEncoder.java:38]: remove unused abstract methods

File: src/checkers/inference/solver/backend/z3smt/encoder/Z3SmtSoftConstraintEncoder.java [hunk 2]
```diff
@@ -34,44 +34,63 @@ public Z3SmtSoftConstraintEncoder(
...
+    protected abstract void encodeSoftSubtypeConstraint(SubtypeConstraint constraint);
+
+    protected abstract void encodeSoftComparableConstraint(ComparableConstraint constraint);
```
Comment (d367wang) [inline src/checkers/inference/solver/backend/z3smt/encoder/Z3SmtSoftConstraintEncoder.java:41]: only reserve subtype, equality, inequality,

File: src/checkers/inference/solver/backend/z3smt/encoder/Z3SmtSoftConstraintEncoder.java [hunk 3]
```diff
@@ -34,45 +35,29 @@ public Z3SmtSoftConstraintEncoder(
     protected void addSoftConstraint(Expr serializedConstraint, int weight) {
         softConstraints.append("(assert-soft " + serializedConstraint + " :weight " + weight + ")\n");
     }
-    
+
```
Comment (zcai1) [inline src/checkers/inference/solver/backend/z3smt/encoder/Z3SmtSoftConstraintEncoder.java:38]:
```suggestion
```

removing an empty line

File: src/checkers/inference/solver/backend/z3smt/encoder/Z3SmtSoftConstraintEncoder.java [hunk 4]
```diff
@@ -34,45 +35,29 @@ public Z3SmtSoftConstraintEncoder(
...
+    protected abstract void encodeSoftSubtypeConstraint(SubtypeConstraint constraint);
+
+    protected abstract void encodeSoftEqualityConstraint(EqualityConstraint constraint);
+
+    protected abstract void encodeSoftInequalityConstraint(InequalityConstraint constraint);
+
     public String encodeAndGetSoftConstraints(Collection<Constraint> constraints) {
         for (Constraint constraint : constraints) {
             // Generate a soft constraint for subtype constraint
```
Comment (d367wang) [inline src/checkers/inference/solver/backend/z3smt/encoder/Z3SmtSoftConstraintEncoder.java:48]: remove comment

File: src/checkers/inference/solver/backend/z3smt/encoder/Z3SmtSoftConstraintEncoder.java [hunk 5]
```diff
@@ -34,45 +35,29 @@ public Z3SmtSoftConstraintEncoder(
...
-            if (constraint instanceof CombineConstraint) {
-                encodeCombineConstraint((CombineConstraint) constraint);
-            }
-            // Generate soft constraint for preference constraint
-            if (constraint instanceof PreferenceConstraint) {
-                encodePreferenceConstraint((PreferenceConstraint) constraint);
+            else if (constraint instanceof InequalityConstraint) {
+                encodeSoftInequalityConstraint((InequalityConstraint) constraint);
             }
+            throw new BugInCF("Soft constraint for " + constraint.getClass().getName() + " is not supported");
```
Comment (d367wang) [inline src/checkers/inference/solver/backend/z3smt/encoder/Z3SmtSoftConstraintEncoder.java:60]: else { throw ...}

File: src/checkers/inference/solver/backend/z3smt/encoder/Z3SmtSoftConstraintEncoder.java [hunk 6]
```diff
@@ -34,45 +35,29 @@ public Z3SmtSoftConstraintEncoder(
...
-    
+
+
+    protected abstract void encodeSoftSubtypeConstraint(SubtypeConstraint constraint);
+
+    protected abstract void encodeSoftEqualityConstraint(EqualityConstraint constraint);
+
+    protected abstract void encodeSoftInequalityConstraint(InequalityConstraint constraint);
+
     public String encodeAndGetSoftConstraints(Collection<Constraint> constraints) {
```
Comment (d367wang) [inline src/checkers/inference/solver/backend/z3smt/encoder/Z3SmtSoftConstraintEncoder.java:46]: reset stringbuilder

File: src/checkers/inference/solver/backend/z3smt/encoder/Z3SmtSoftConstraintEncoder.java [hunk 7]
```diff
@@ -34,44 +35,35 @@ public Z3SmtSoftConstraintEncoder(
...
+
+    protected abstract void encodeSoftSubtypeConstraint(SubtypeConstraint constraint);
+
+    protected abstract void encodeSoftEqualityConstraint(EqualityConstraint constraint);
+
+    protected abstract void encodeSoftInequalityConstraint(InequalityConstraint constraint);
+
+    /**
+     * Encode a set of constraints as soft constraints. Note that the field of StringBuilder
+     * {@code softConstraints} is first cleared each time this method is called.
```
Comment (wmdietl) [inline src/checkers/inference/solver/backend/z3smt/encoder/Z3SmtSoftConstraintEncoder.java:47]:
```suggestion
     * Encode a set of constraints as soft constraints.
     *
```

File: src/checkers/inference/solver/backend/z3smt/encoder/Z3SmtSoftConstraintEncoder.java [hunk 8]
```diff
@@ -34,44 +35,35 @@ public Z3SmtSoftConstraintEncoder(
...
+                encodeSoftEqualityConstraint((EqualityConstraint) constraint);
+
+            } else if (constraint instanceof InequalityConstraint) {
+                encodeSoftInequalityConstraint((InequalityConstraint) constraint);
+
+            } else {
+                throw new BugInCF("Soft constraint for " + constraint.getClass().getName() + " is not supported");
             }
         }
         return softConstraints.toString();
```
Comment (wmdietl) [inline src/checkers/inference/solver/backend/z3smt/encoder/Z3SmtSoftConstraintEncoder.java:69]:
```suggestion
        String res = softConstraints.toString();
        // clear field for next usage
        softConstraints.setLength(0);
        return res;
```

File: src/checkers/inference/solver/backend/z3smt/encoder/Z3SmtSoftConstraintEncoder.java [hunk 9]
```diff
@@ -34,44 +35,35 @@ public Z3SmtSoftConstraintEncoder(
...
+    /**
+     * Encode a set of constraints as soft constraints. Note that the field of StringBuilder
+     * {@code softConstraints} is first cleared each time this method is called.
+     * @param constraints constraints to be encoded as soft constraints
+     * @return a string representation of the encoding of soft constraints
+     */
     public String encodeAndGetSoftConstraints(Collection<Constraint> constraints) {
+        // clear previous encoding result before encoding
+        softConstraints.setLength(0);
+
```
Comment (wmdietl) [inline src/checkers/inference/solver/backend/z3smt/encoder/Z3SmtSoftConstraintEncoder.java:54]:
```suggestion
```

**Pull request merged**